scipy/stats/_hypotests.py

```diff
@@ -1433,3 +1435,305 @@ def cramervonmises_2samp(x, y, method='auto'):
+           The ANOVA table and tests of hypotheses about means"
+           https://www.itl.nist.gov/div898/handbook/prc/section4/prc433.html,
+           2 June 2021.
+    .. [6] Tukey, John W. "Comparing Individual Means in the Analysis of
+           Variance." Biometrics, vol. 5, no. 2, 1949, pp. 99-114. JSTOR,
+           www.jstor.org/stable/3001913. Accessed 14 June 2021.
+
+
+    Examples
+    --------
+    Here are some data comparing the time to relief of three brands of
+    headache medicine, reported in minutes. Data adapted from [3]_.
+
+    >>> from scipy.stats import tukey_hsd
+    >>> group0 = [24.5, 23.5, 26.4, 27.1, 29.9]
+    >>> group1 = [28.4, 34.2, 29.5, 32.2, 30.1]
+    >>> group2 = [26.1, 28.3, 24.3, 26.2, 27.8]
+
+    We would like to see if the means between any of the groups are
+    significantly different. First, visually examine a box and whisker plot.
+
+    >>> import matplotlib.pyplot as plt
+    >>> fig, ax = plt.subplots(1, 1)
+    >>> ax.boxplot([group0, group1, group2])
+    >>> ax.set_xticklabels(["group0", "group1", "group2"]) # doctest: +SKIP
+    >>> ax.set_ylabel("mean") # doctest: +SKIP
+    >>> plt.show()
+
+    From the box and whisker plot we can see overlap in the interquartile
+    ranges group 1 to group 2 and group 3, but we can apply the ``tukey_hsd``
+    test to determine if the difference between means is significant. We
+    a significance level of .05 to reject the null hypothesis.
+
+    >>> res = tukey_hsd(group0, group1, group2)
+    >>> print(res)
+    Tukey's HSD Pairwise Group Comparisons
+    Comparison  Statistic  p-value
+    (0 - 0)      0.000      1.000
+    (0 - 1)     -4.600      0.014
+    (0 - 2)     -0.260      0.980
+    (1 - 0)      4.600      0.014
+    (1 - 1)      0.000      1.000
+    (1 - 2)      4.340      0.020
+    (2 - 0)      0.260      0.980
+    (2 - 1)     -4.340      0.020
+    (2 - 2)      0.000      1.000
+
+    The null hypothesis is that each group has the same mean. The p-value for
+    comparisons between ``group0`` and ``group1`` as well as ``group1`` and
+    ``group2`` do not exceed .05, so we reject the null hypothesis that they
+    have the same means. The p-value of the comparison between ``group0``
+    and ``group2`` exceeds .05, so we accept the null hypothesis that there
+    is not a significant difference between their means.
+
+    We can also compute the confidence interval associated with our chosen
+    confidence level.
+
+    >>> group0 = [24.5, 23.5, 26.4, 27.1, 29.9]
+    >>> group1 = [28.4, 34.2, 29.5, 32.2, 30.1]
+    >>> group2 = [26.1, 28.3, 24.3, 26.2, 27.8]
+    >>> result = tukey_hsd(group0, group1, group2)
+    >>> conf = res.confidence_interval()
+    >>> for ((i, j), l) in np.ndenumerate(conf.low):
+    ...     # filter out self comparisons
+    ...     if i != j:
+    ...         h = conf.high[i,j]
+    ...         print(f"({i} - {j}) {l:>6.3f} {h:>6.3f}")
```

Comment:
This looks like something that a user might want to do if they compute the confidence intervals no?

What about returning the data like that, but also storing the results in the result class? This way we could adapt the `__str__` to include or not the confidence interval. This could even be asked by the user directly with `tukey_hsd(groups: array_like, confidence_interval: Optional[float])`

Reply:
I think you're suggesting that the user might want to calculate all the confidence intervals? I think the reason they are not calculated by default is to try to be consistent with other functions that defer CI calculation until the `confidence_interval` method of the result is called. I'm conflicted, because you're right @tupui we'd want the table printed by `__str__`  to include the CIs (like every other implementation of Tukey's HSD).

Reply:
We can still not compute CI by default: we can have an if in `__str__` which does something different if CI was previously asked or not. This would work even without adding an option to the function. But the design would be a bit weird as one could call `__str__`, then ask for CI and call again  `__str__` which would produce something different.

Reply:
I agree that the possibility of calling `__str__` before and after `confidence_interval()` would give different outputs could cause confusion. 

Maybe moving the dynamic-ness to the `confidence_interval()` method would be better. What if there was a private variable in the class that caches the confidence interval, so if the user calls `__str__` the CIs are generated and stored, and then calling `confidence_interva()` separately would check to see if CIs had already been generated?

Then CIs are still only calculated upon user request, as getting the string representation is a type of request.

Reply:
Do I understand correctly that you propose calling `__str__` would always return CI? Fine with me. As for the caching, it's a good idea and we can just go with a private attribute in the class and an if (lru cache and others would be overkill I think).

Reply:
Yeah, actually. It's a little weird, but I think it meets the goals.

Reply:
I supposed that a follow up complication to this would be from which confidence level the confidence interval `__str__` should then display:
- it could call the `confidence_interval()` method, either getting the default confidence level of .95 or the most recently calculated (and cached with a private variable) ci with whatever confidence level the user specified. This would mean that if the user printed the result before calling any confidence interval method, it would display the confidence intervals for confidence level .95, but then if the user calculated the confidence intervals for .99 and then printed the result, it would be different.
- it could call the confidence interval method always with .95 (the default confidence level for the method). If .95 is the confidence level for the confidence intervals that were cached then it will use that, otherwise calculate and re-cache.

What I am weighing is how worth it it is to have the `__str__` output be constant. I think there is the potential to be frustrating to print the result multiple times and get different results. In either scenario it would be wise to indicate the confidence level for the confidence intervals in the printout, as I too want `__str__` to always return CIs.

Reply:
I like your first option with the addition of the confidence level value in the `__str__`. This would be clear then and to me it would be fine then to have a mutable method output. In the end it would correspond better to the "current state" of the object.